Added new effect 'Random Flash'

Lukiolio opened the pull request:
Disclaimer: This is my first contribution ever to an open source software so please be gentle :) I am glad for any kind of constructive feedback!

I added a new effect called "random" which randomly activates a section of leds based on the params "hit duration", "hit probability per second" and "relative hit size" (relative to whole strip size). I was inspired by a thunderstorm-like scene with random flashes.
![Bildschirmfoto 2024-10-13 um 18 35 34](https://github.com/user-attachments/assets/dc1bf9df-7e43-4ac7-aed8-9b80d9680210)
https://github.com/user-attachments/assets/44cd6f74-81ba-4554-93ee-7ec947c79370




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new lighting effect called "Random" that randomly activates sections of an LED strip.
	- Added a "Random Flash" effect that simulates a lightning-like appearance on the LED strip.
	- Allows customization of hit color, duration, probability, and size for enhanced user control over the effects.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,92 @@
+                "hit_relative_size",
+                description="Hit size relative to LED strip",
+                default=0.2,
+            ): vol.All(vol.Coerce(float), vol.Range(min=0.1, max=1.0)),
```
You might want to consider a minimum size smaller than 0.1, that is 1 tenth of the overall strip length.

You could consider making this slider to a percentage int, range 1 to 100

Its not consistant in the code base, but for example in the scan effect, I do that in the scan effect

https://github.com/LedFx/LedFx/blob/main/ledfx/effects/scan.py#L44

There are likely example of both styles

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,92 @@
+        hit_absolute_size = int(self.pixel_count * self.hit_relative_size)
+        self.__update_duration_since_last_hit()
+
+        hit_is_still_active = self.seconds_since_last_hit < self.hit_duration
```
This should just be the if statement clause, doesn't need to be substantiated as a variable.

if not ( self.seconds_since_last_hit < self.hit_duration)

if similar

Lukiolio replied:
I thought it would add to the readability, although it might be obsolete

bigredfrog replied:
It makes you expect to see that used as a variable elsewhere, better to have a comment above to explain the logic if it seems hard to read. This is not a complex comparison, but comments don't hurt.

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,92 @@
+
+    def __update_duration_since_last_hit(self):
+        # NOTE: speed 0.1 is 1 update per sec
+        self.seconds_since_last_hit += 1 / (10 * self.speed)
```
Is this seconds since last hit, or actually some form of counting frames.

Read actual realtime with something like 

        now = timeit.default_timer()
        time_passed = now - self.last_time

To give you frame indepenance.

bigredfrog replied:
Also consider inplementing a hold and decay time. Lightning like effect will look better if it fades out.

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,92 @@
+                # frame slice based of random_pos will be hited
+                frame[random_pos : random_pos + hit_absolute_size, :] = (
+                    np.tile(self.hit_color, (hit_absolute_size, 1))
+                )
```
This looks over complicated, should just be able to do something like

frame[random_pos : random_pos + hit_absolute_size] = self.hit_color

That should copy the color into every array index in range

I would also consider writing all of this directly into self.pixels and save yourself a copy...

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,92 @@
+import random
+import time
```
This is what is making your precommit checks fail. 

remove time reference

However, I suspect you should add timeit and implement time invariance.

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,91 @@
+
+    NAME = "Random"
+    CATEGORY = "Non-Reactive"
+
```
add a hidden key for "flip" as it won't do anything for your effect, it has no directionality, so just hide it.

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,91 @@
+    Randomly activates sections
+    """
+
+    NAME = "Random"
```
I would recommend a name change to something like lightning, more descriptive of intent. Random is going to be interpreted to widely.

Changing the name is as simple as changing the file name, class name and this NAME, its easier than it sounds.

I still would recommend looking into implementing a decay on the "strike" which is as simple as a float 1->0 decay multiplier of the color array. It will look a lot more organic.

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,91 @@
+        self.speed = self._config["speed"]
+        self.probability_per_sec = (
+            self.__balance_hit_probability_based_on_speed()
+        )
```
why the scope brackets?

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,91 @@
+    def __balance_hit_probability_based_on_speed(self) -> float:
+        runs_per_sec = self.speed * 10
+        # this is the probability per effect run
+        return 1 - (1 - self._config["hit_probability_per_sec"]) ** (
```
What is the difference between speed and hit probability per sec, shouldn't this just be one param?

Lukiolio replied:
I find it to be rather unintuitive to use the speed param to model the hit probability. Of course it's possible though. Do you have some adaption in mind how to map the speed param to a hit probability?

bigredfrog replied:
I think this is just a case of semantics. If both sliders are only used for this one thing, then it only needs a single slide with a good name.

Speed just looks like a multiplier against probability. So end user shouldn't have two sliders to care about.

I need to look at this from a PC again, I might be missing something from phone

bigredfrog commented:
I am likely to keep commenting past the horizon of usefulness, however, I would look to at least change the name, and look closely at speed + hit probability per sec.

A user only really wants one slider for one thing ( don't make them do math )

bigredfrog commented on ledfx/effects/random.py:
```diff
@@ -0,0 +1,91 @@
+                    self.pixel_count - hit_absolute_size + 1
+                )
+                # frame slice based of random_pos will be hit
+                self.pixels[random_pos : random_pos + hit_absolute_size] = (
```
spurious scope brackets

Lukiolio commented:
I renamed the effect, removed the speed param and added a fade to the flash (although I am still not 100% sure about my implementation of the latter). By the way, the bracket placement is determined by the pre-commit bot so I don't really have any influence of that

bigredfrog commented on ledfx/effects/random_flash.py:
```diff
@@ -0,0 +1,101 @@
+                self.last_time = timeit.default_timer()
+
+    def __balance_hit_probability_based_on_speed(self) -> float:
+        runs_per_sec = self._config["speed"] * 10
```
You can get rid of this line now, and just use that runs per sec as a constant, you don't want to be referencing a var unnecessarily.

Lukiolio replied:
Good point, fixed it

bigredfrog commented on ledfx/effects/random_flash.py:
```diff
@@ -0,0 +1,101 @@
+                    self.pixel_count - hit_absolute_size + 1
+                )
+                # frame slice based of random_pos will be hit
+                self.pixels[random_pos : random_pos + hit_absolute_size] = (
```
This just needs to be

self.pixels[random_pos : random_pos + hit_absolute_size] = self.hit_color

bigredfrog commented:
We are into "I would of done it different" territory now, so do you want to know how I would of sliced this, or do you just want to get it in for now?

bigredfrog commented on ledfx/effects/random_flash.py:
```diff
@@ -0,0 +1,102 @@
+        self.last_time = timeit.default_timer()
+
+    def effect_loop(self):
+        hit_absolute_size = int(
```
move this to self.hit_absolute_size and calculate in the config_updated()

No need to be recalculating this every render pass

bigredfrog replied:
ah, of course, I forget that sometimes that pixel_count is not available in first cycle throgh config_updated()

Will resolve

bigredfrog commented on ledfx/effects/random_flash.py:
```diff
@@ -0,0 +1,102 @@
+        self.hit_duration = self._config["hit_duration"]
+        self.probability_per_sec = (
+            self.__balance_hit_probability_based_on_speed()
+        )
```
I mean why have the brackets at all, did the syntax checker add them?

Lukiolio replied:
Yep, that was not my work.

Lukiolio commented:
I am always happy to see a different approach :) However, I also think that this kind of "I would do it differently" could go on forever, so I am glad that the major and some minor flaws are fixed now and I personally would call it a day now @bigredfrog, are you fine with that?

bigredfrog commented:
Tested locally. Merging.

Remember you can iterate forward at any time...

shauneccles commented:
Welcome to the codebase @Lukiolio it's nice to have you 🥳

Lukiolio commented:
Thank you! :)
